CDS-49944 (#417)
harness/developer-hub: docs/continuous-delivery/onboard-cd/upgrading/cdng-upgrade-faq.md

---
title: Harness CD upgrading FAQ
description: Answers to common upgrade questions.
sidebar_position: 1
---

This topic provides answers to common questions about upgrading from Harness Continuous Delivery (CD) FirstGen to NextGen.

For steps on upgrading, go to [Upgrade to CD NextGen](upgrade-nextgen-cd.md).

For information on the differences between Harness FirstGen and NextGen, go to [Harness FirstGen vs Harness NextGen](../../../getting-started/harness-first-gen-vs-harness-next-gen.md).

## What is Harness Continuous Delivery NextGen?

Harness has reimagined Continuous Delivery (CD) and built a new version of our CD product. 

Continuous Delivery is now part of a broader platform known as the Harness Software Delivery Platform. 

> We intend for the platform to improve the developer experience. 

Reviewer: Do we use "We"? I thought we use only "Harness"

Author: We is allowed per MS Style Guide:
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/w/we

> It's OK to use phrasing like we recommend if it helps you avoid awkward phrasing like it's recommended, but write around it if you can.

### Enterprise GitOps

NextGen includes the following GitOps benefits:

- Harness brings enterprise controls to GitOps deployment with governance, reliability, and visibility at scale.
- Harness GitOps delivers a fully managed GitOps experience that allows teams to scale implementations quickly and spend less time managing environments.

### Continuous Verification (CV) NextGen

Harness CV integrates with your APMs and logging tools to:

- Verify that the deployed service is running safely and perform automatic rollbacks.
- Apply machine learning to every deployment to identify and flag anomalies in future deployments.

Some CV features:

- Harness includes a **Verify** step that queries logging and metrics providers from multiple sources at once.
- Next, Harness uses its AI/ML-driven verification capability to automatically validate deployment quality. This capability reduces risk and provides a safety net when code is deployed to production.
- Integrations include Prometheus, Splunk, Google Cloud Operations, Datadog, and other leading logging and metric providers.

### Policy as code

> NextGen provides the following governance policy benefits:

Author: added

  - For details on NextGen SSO providers, go to [Authentication overview](https://developer.harness.io/docs/platform/Authentication/authentication-overview).
- **Deployment history for CD dashboards**. Harness NextGen does not retain the deployment data from FirstGen. Harness will recalculate the deployment stats and metrics using NextGen CD deployments.

  - For details on NextGen dashboards, go to [Monitor deployments and services in CD dashboards](https://developer.harness.io/docs/continuous-delivery/cd-dashboards/monitor-cd-deployments).
- **API Automation**. Harness will not upgrade API automation built with FirstGen GraphQL APIs.
    
    GraphQL APIs are not used in NextGen. They are replaced with REST APIs. The API endpoints are different and take different arguments. You will need to rewrite API-based automation to reintegrate with Harness NextGen.

  - For details on NextGen APIs, go to [APIs](https://developer.harness.io/docs/category/apis).
- **Artifact Collection**. Harness will not migrate the artifact history of a FirstGen artifact source to NextGen.
    

>     Harness NextGen does not perform artifact polling to collect the list of artifacts and maintain a history for future selection. NextGen now fetches the list of tags at pipeline runtime. 

Reviewer 2: ```suggestion
    Harness NextGen does not perform artifact polling to collect the list of artifacts and maintain a history for future selection. NextGen fetches the list of tags at pipeline runtime. 
```
(The "now" implies that NextGen did this differently previously. Did it? That is not coming out in this text. If no, removing "now" helps.)

  - For details on NextGen artifact collection, review the artifact sources covered in [Harness Kubernetes services](https://developer.harness.io/docs/continuous-delivery/cd-services/k8s-services/kubernetes-services).

## What are my CD NextGen upgrade options?

There are two Harness CD NextGen upgrade options.

> ### Option 1: Harness assisted upgrade (recommended)

Reviewer 2: ```suggestion
### Option 1: Harness-assisted upgrade (recommended)
```

> 1. Customers schedule a meeting with the Harness team to help plan for an automated upgrade.
> 2. Harness reviews the CD FirstGen account and goes over a proposed upgrade plan with customers.
> 3. Harness upgrades the configuration to CD NextGen using our upgrade APIs. This process will not impact your CD FirstGen setup.
> 4. Customers and the Harness team review the upgraded configuration.
> 5. Customers build their new CD NextGen pipelines and validate the deployed apps through those pipelines. Harness provides guidance as needed.
> 6. Customers and the Harness team work together to ensure a smooth upgrade and hit the agreed upon upgrade timeline.

Reviewer 2: ```suggestion
1. You schedule a meeting with the Harness team to help plan for an automated upgrade.
2. Harness reviews the CD FirstGen account and proposes an upgrade plan to you.
3. Harness upgrades the configuration to CD NextGen by using its upgrade APIs. This process will not impact your CD FirstGen setup.
4. You and the Harness team review the upgraded configuration.
5. You build your new CD NextGen pipelines and validate the apps deployed through those pipelines. Harness provides guidance as needed.
6. You and the Harness team work together to ensure a smooth upgrade and meet the agreed upon upgrade timeline.
```
(Good to use "you" instead of "customer". Also, the question uses "my", so...)

> #### Timeline

Reviewer 2: This is the only level-4 heading under option 1. Can we drop the heading and integrate this content with step 6 above? For example:
~~
6. You and the Harness team work together to ensure a smooth upgrade and meet the agreed upon upgrade timeline. With the Harness upgrade team involved, the estimated timeline for a complete upgrade is about six weeks. Here is the upgrade process and the timeline for each stage.
![picture 1](static/362da5ad5713ca1d3a30786f24eb71f6ce60ec094a00abf7e93b02bb8c2e0422.png)
~~
Ignore this suggestion if you need to call out the timeline content for emphasis or make it easy to pick out when visually scanning.

> With the Harness upgrade team involved, the estimated timeline for a complete upgrade is about 6 weeks. Here is the upgrade process and the timeline per each stage.

Reviewer 2: ```suggestion
With the Harness upgrade team involved, the estimated timeline for a complete upgrade is about six weeks. Here is the upgrade process and the timeline for each stage.
```
(Per MMSTP at https://learn.microsoft.com/en-us/style-guide/numbers)

![picture 1](static/362da5ad5713ca1d3a30786f24eb71f6ce60ec094a00abf7e93b02bb8c2e0422.png)

### Option 2: Self-service upgrade

> 1. Customers access Harness CD NextGen from the homepage of Harness FirstGen.
> 
>     ![picture 2](static/d70937e8aa1a50af75650d5a64bfdb0e9e2bfc2278ff6fb7d4c5019b1487ef17.png)  
> 
> 2. Customers pick the FirstGen applications to recreate in NextGen.
>    
>    This includes the connectors, delegates, services, environments, workflows, and pipelines.
> 3. Customers deploy and test the new pipeline(s) in Harness CD NextGen.
> 4. Once customers are comfortable deploying with Harness CD NextGen, they can remove the application from FirstGen.
> 
> Customers can reach out to Harness for review and validation of their Harness CD NextGen pipelines.

Reviewer 2: ```suggestion
1. You access Harness CD NextGen from the home page of Harness FirstGen.

    ![picture 2](static/d70937e8aa1a50af75650d5a64bfdb0e9e2bfc2278ff6fb7d4c5019b1487ef17.png)  

2. You pick the FirstGen applications to re-create in NextGen.
   
   This includes connectors, delegates, services, environments, workflows, and pipelines.
3. You deploy and test the new pipelines in Harness CD NextGen.
4. When you are comfortable deploying with Harness CD NextGen, you can remove the application from FirstGen.

You can reach out to Harness for review and validation of your Harness CD NextGen pipelines.
```

#### Timeline 

> Harness customers share the start date and end date for the upgrade, including the final cutover date. This enables the Harness upgrade team to provide any assistance needed during customers' self-service upgrade.

Reviewer 2: ```suggestion
You share the start date and end date for the upgrade, including the final cutover date. This enables the Harness upgrade team to provide any assistance needed during your self-service upgrade.
```

## Are all features from FirstGen available in NextGen?

Yes. NextGen has feature parity. 

Some deployment capabilities have been overhauled:
- AWS ECS deployments have a significantly different experience from FirstGen.

> - Deployment templates has been extended. 

Reviewer 2: "have been extended" for subject-verb agreement? Or did you mean "The deployment templates feature has been extended"?

Author: should be "have". thx

- Traditional (SSH and WinRM) deployments have a new configuration experience.

> - FirstGen Infrastructure Provisioners are now NextGen provisioning steps. You can manage the entire infrastructure provisioning configuration in each step.

Reviewer 2: ```suggestion
- Infrastructure provisioners in FirstGen are now provisioning steps in NextGen. You can manage the entire infrastructure provisioning configuration in each step.
```

> ## What major benefits have upgraded customers noticed in NextGen?

Reviewer 2: Could improve parallelism in the items here but that's a nit.

- Full config-as-code. Git is the source of truth for your pipelines and templates.
- Significant efficiencies in build and delivery lead times.
- OPA-based governance helps enforce policies in pipeline designs.
- Standardized deployment processes using shareable templates.
- Reduction in workflow and pipeline management as more common components can be shared.

> - Hybrid (multi-cloud deployment) pipelines and ad-hoc jobs are easier to configure and manage.

Reviewer 2: ```suggestion
- Hybrid (multicloud deployment) pipelines and ad-hoc jobs are easier to configure and manage.
```
(https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/m/multi)

> - Multi-service and multi-environment deployments at scale.

Reviewer 2: https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/m/multi recommends not using a hyphen with multi unless using a hyphen helps avoid confusion or multi precedes a proper noun. It recommends using "multiple" if the word we are looking for help with isn't on the list on that page. I'm okay leaving this in its current state if there isn't a more elegant way to reword this. :)
